File: code/client/core/source/operation-graph/OperationGraphWriter.cpp
﻿// <copyright file="OperationGraphWriter.cpp" company="Soup">
// Copyright (c) Soup. All rights reserved.
// </copyright>

module;

#include <cstdint>
#include <cstring>
#include <iostream>
#include <set>
#include <vector>

export module Soup.Core:OperationGraphWriter;

import Opal;
import :FileSystemState;
import :OperationGraph;
import :OperationInfo;

using namespace Opal;

namespace Soup::Core
{
	/// <summary>
	/// The operation graph state writer
	/// </summary>
	export class OperationGraphWriter
	{
	private:
		// Binary Operation graph file format
		static constexpr uint32_t FileVersion = 6;

	public:
		static void Serialize(
			const OperationGraph& state,
			const std::set<FileId>& files,
			const FileSystemState& fileSystemState,
			std::ostream& stream)
		{
			// Write the File Header with version
			stream.write("BOG\0", 4);
			WriteValue(stream, FileVersion);

			// Write out the set of files
			stream.write("FIS\0", 4);
			WriteValue(stream, static_cast<uint32_t>(files.size()));
			for (auto fileId : files)
			{
				// Write the file id + path length + path
				WriteValue(stream, fileId);
				WriteValue(stream, fileSystemState.GetFilePath(fileId).ToString());
			}

			// Write out the root operation ids
			stream.write("ROP\0", 4);
			WriteValues(stream, state.GetRootOperationIds());

			// Write out the set of operations
			auto& operations = state.GetOperations();
			stream.write("OPS\0", 4);
			WriteValue(stream, static_cast<uint32_t>(operations.size()));
			for (auto& operationValue : operations)
			{
				WriteOperationInfo(stream, operationValue.second);
			}
		}

		static void WriteOperationInfo(std::ostream& stream, const OperationInfo& operation)
		{
			// Write out the operation id
			WriteValue(stream, operation.Id);

			// Write out the operation title
			WriteValue(stream, operation.Title);

			// Write the command working directory
			WriteValue(stream, operation.Command.WorkingDirectory.ToString());

			// Write the command executable
			WriteValue(stream, operation.Command.Executable.ToString());

			// Write the command arguments
			WriteValues(stream, operation.Command.Arguments);

			// Write out the declared input files
			WriteValues(stream, operation.DeclaredInput);

			// Write out the declared output files
			WriteValues(stream, operation.DeclaredOutput);

			// Write out the read access list
			WriteValues(stream, operation.ReadAccess);

			// Write out the write access list
			WriteValues(stream, operation.WriteAccess);

			// Write out the child operation ids
			WriteValues(stream, operation.Children);

			// Write out the dependency count
			WriteValue(stream, operation.DependencyCount);
		}

	private:
		static void WriteValue(std::ostream& stream, uint32_t value)
		{
			stream.write(reinterpret_cast<char*>(&value), sizeof(uint32_t));
		}

		static void WriteValue(std::ostream& stream, int64_t value)
		{
			stream.write(reinterpret_cast<char*>(&value), sizeof(int64_t));
		}

		static void WriteValue(std::ostream& stream, bool value)
		{
			uint32_t integerValue = value ? 1u : 0u;
			stream.write(reinterpret_cast<char*>(&integerValue), sizeof(uint32_t));
		}

		static void WriteValue(std::ostream& stream, std::string_view value)
		{
			WriteValue(stream, static_cast<uint32_t>(value.size()));
			stream.write(value.data(), value.size());
		}

		static void WriteValues(std::ostream& stream, const std::vector<std::string>& values)
		{
			WriteValue(stream, static_cast<uint32_t>(values.size()));
			for (auto& value : values)
			{
				WriteValue(stream, value);
			}
		}

		static void WriteValues(std::ostream& stream, const std::vector<uint32_t>& values)
		{
			WriteValue(stream, static_cast<uint32_t>(values.size()));
			for (auto& value : values)
			{
				WriteValue(stream, value);
			}
		}
	};
}

File: code/generate-sharp/utilities/operation-graph/operation-graph-writer.cs
﻿// <copyright file="operation-graph-writer.cs" company="Soup">
// Copyright (c) Soup. All rights reserved.
// </copyright>

using System.Collections.Generic;
using System.IO;

namespace Soup.Build.Utilities;

/// <summary>
/// The operation graph state writer
/// </summary>
internal static class OperationGraphWriter
{
	// Binary Operation graph file format
	private static uint FileVersion => 6;

	internal static readonly char[] FIS = ['F', 'I', 'S', '\0'];
	internal static readonly char[] BOG = ['B', 'O', 'G', '\0'];
	internal static readonly char[] ROP = ['R', 'O', 'P', '\0'];
	internal static readonly char[] OPS = ['O', 'P', 'S', '\0'];

	public static void Serialize(OperationGraph state, BinaryWriter writer)
	{
		// Write the File Header with version
		writer.Write(BOG);
		writer.Write(FileVersion);

		// Write out the set of files
		var files = state.ReferencedFiles;
		writer.Write(FIS);
		writer.Write((uint)files.Count);
		foreach (var file in files)
		{
			// Write the file id + path length + path
			writer.Write(file.FileId.Value);
			WriteValue(writer, file.Path.ToString());
		}

		// Write out the root operation ids
		writer.Write(ROP);
		WriteValues(writer, state.RootOperationIds);

		// Write out the set of operations
		writer.Write(OPS);
		writer.Write((uint)state.Operations.Count);
		foreach (var operationValue in state.Operations)
		{
			WriteOperationInfo(writer, operationValue.Value);
		}
	}

	public static void WriteOperationInfo(BinaryWriter writer, OperationInfo operation)
	{
		// Write out the operation id
		writer.Write(operation.Id.Value);

		// Write out the operation title
		WriteValue(writer, operation.Title);

		// Write the command working directory
		WriteValue(writer, operation.Command.WorkingDirectory.ToString());

		// Write the command executable
		WriteValue(writer, operation.Command.Executable.ToString());

		// Write the command arguments
		WriteValues(writer, operation.Command.Arguments);

		// Write out the declared input files
		WriteValues(writer, operation.DeclaredInput);

		// Write out the declared output files
		WriteValues(writer, operation.DeclaredOutput);

		// Write out the read access list
		WriteValues(writer, operation.ReadAccess);

		// Write out the write access list
		WriteValues(writer, operation.WriteAccess);

		// Write out the child operation ids
		WriteValues(writer, operation.Children);

		// Write out the dependency count
		writer.Write(operation.DependencyCount);
	}

	private static void WriteValue(BinaryWriter writer, string value)
	{
		writer.Write((uint)value.Length);
		writer.Write(value.ToCharArray());
	}

	private static void WriteValues(BinaryWriter writer, IReadOnlyList<string> values)
	{
		writer.Write((uint)values.Count);
		foreach (var value in values)
		{
			writer.Write(value);
		}
	}

	private static void WriteValues(BinaryWriter writer, IList<FileId> values)
	{
		writer.Write((uint)values.Count);
		foreach (var value in values)
		{
			writer.Write(value.Value);
		}
	}

	private static void WriteValues(BinaryWriter writer, IList<OperationId> values)
	{
		writer.Write((uint)values.Count);
		foreach (var value in values)
		{
			writer.Write(value.Value);
		}
	}
}